### PR TITLE
Selectable directive

### PR DESCRIPTION
**Pull request by angrykoala**
# Description
Adds `@selectable` directive defined in #3263 

This directive has the following options:
* onRead
* onAggregate

Setting these values will disable the generation of the GraphQL fields for responses or aggregations

## Complexity

Complexity: Medium

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 35 | 39 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 68 | 107 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 48 | 79 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3150 | 3073 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2893 | 3108 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 82 | 47 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 74 | 48 | 2174 |
|  aggregations.DeeplyNestedCount | 13070329 | 13070331 | 4431 | 4534 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 124 | 134 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 84 | 60 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 175 | 200 | 3003 |
|  connections.Connection | 14082 | 14082 | 68 | 75 | 2174 |
|  connections.NestedConnection | 45466 | 45470 | 109 | 135 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 64 | 69 | 1040 |
|  connections.ConnectionWithSortAndCypher | 16325 | 16325 | 86 | 99 | 2174 |
|  create.SimpleMutation | 7 | 7 | 51 | 59 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 29 | 62 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 257 | 298 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 203 | 177 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 68 | 45 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 123 | 104 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 48 | 44 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 55 | 70 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 24 | 30 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 41 | 50 | 2174 |
|  query.QueryWhere | 9690 | 9715 | 36 | 40 | 2160 |
|  query.SimpleQueryWithNestedWhere | 9868 | 9893 | 55 | 58 | 2160 |
|  query.Nested | 10096041 | 10096041 | 10313 | 10113 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11078 | 10095 | 2004000 |
|  query.OrFilterOnRelationships | 43074 | 41363 | 195 | 187 | 1781 |
|  query.OrFilterOnRelationshipsAndNested | 36803 | 35501 | 216 | 257 | 1781 |
|  query.QueryWithNestedIn | 13285 | 13368 | 60 | 59 | 1177 |
|  query.NestedConnectionWhere | 9834 | 9834 | 54 | 63 | 2174 |
|  query.DeeplyNestedConnectionWhere | 10104 | 9984 | 73 | 117 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19326 | 19164 | 126 | 233 | 1658 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 153 | 371 | 1134 |
|  query.Fulltext | 80 | 80 | 30 | 36 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 43 | 53 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 64 | 73 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14512 | 14501 | 86 | 110 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 50 | 55 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 87 | 104 | 4516 |
|  sorting.SortWithTopLevelCypher | 16162 | 16162 | 56 | 54 | 2174 |
|  unions.SimpleUnionQuery | 321 | 321 | 60 | 65 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 63 | 75 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 380 | 327 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 295 | 353 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 94 | 197 | 2002 |

</details>
        
Old Schema Generation: 38.436s
Schema Generation: 36.588s
Old Subgraph Schema Generation: 50.168s
Subgraph Schema Generation: 48.224s

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+
+    test("Disable read fields", async () => {
+        const typeDefs = gql`
+            type Movie {
+                title: String!
+                description: String @selectable(onRead: false, onAggregate: true)
+            }
+        `;
+        const neoSchema = new Neo4jGraphQL({ typeDefs });
+        const printedSchema = printSchemaWithDirectives(lexicographicSortSchema(await neoSchema.getSchema()));
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:41]: Is it worth updating the tests to use the new style of assertions for schema tests such as here? https://github.com/neo4j/graphql/blob/02909bb18fa482a196072d24c3ed39080dbb2e68/packages/graphql/tests/schema/directives/query.test.ts#L26

File: packages/graphql/src/graphql/directives/selectable.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DirectiveLocation, GraphQLBoolean, GraphQLDirective, GraphQLNonNull } from "graphql";
+
+export const selectableDirective = new GraphQLDirective({
+    name: "selectable",
+    description: "Instructs @neo4j/graphql to generate this field for read operations.",
```
[the remark this answers is not in the document]
Comment (angrykoala) [reply packages/graphql/src/graphql/directives/selectable.ts:24]: mmm, `types`? Isn't `selectable fields` more accurate maybe?

[the remark this answers is not in the document]
Comment (Liam-Doodson) [reply packages/graphql/src/graphql/directives/selectable.ts:24]: Is it not type?

Surely this is saying generate the `title` field on the `Movie` type:

```gql
type Movie {
    title: String @selectable
}
```

[the remark this answers is not in the document]
Comment (angrykoala) [reply packages/graphql/src/graphql/directives/selectable.ts:24]: Oh, right, I misread

File: packages/graphql/tests/schema/directives/selectable.test.ts (hunk 1 above)
Comment (angrykoala) [reply packages/graphql/tests/schema/directives/selectable.test.ts:41]: It is definitely worth it at some point, however, the approach taken in `query.test.ts` is quite naive, as it only checks for top-level existence of fields, while this test needs to check for fields. Parsing the graphql tree in our tests directly is going to be a recipe for complex and hard testing. We need to find a better approach for this IMO. So if that's ok, I'll leave these tests as they are and make a point to discuss this in the future

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+describe("@selectable", () => {
+    let subscriptionPlugin: TestSubscriptionsPlugin;
+
+    beforeAll(() => {
+        subscriptionPlugin = new TestSubscriptionsPlugin();
+    });
+
+    test("Disable read fields", async () => {
+        const typeDefs = gql`
+            type Movie {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:35]: Is it worth explicitly enabling subscriptions/aggregations for these tests in preparation for the 4.0.0 changes?
```suggestion
            type Movie @query(aggregate: true) {
```

**Review by Liam-Doodson — DISMISSED** [review]
LGTM!

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 4]
```diff
@@ -0,0 +1,1611 @@
...
+            type UpdateMoviesMutationResponse {
+              info: UpdateInfo!
+              movies: [Movie!]!
+            }"
+        `);
+    });
+
+    test("Disable aggregation fields", async () => {
+        const typeDefs = gql`
+            type Movie {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:193]: As above, I think we should explicitly enable aggregations here so we don't lose the parts of the schema we are interested in when we change the default behaviour

```suggestion
            type Movie @query(aggregate: true)  {
```

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 5]
```diff
@@ -0,0 +1,1611 @@
...
+            type UpdateMoviesMutationResponse {
+              info: UpdateInfo!
+              movies: [Movie!]!
+            }"
+        `);
+    });
+
+    test("Disable read and aggregate fields", async () => {
+        const typeDefs = gql`
+            type Movie {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:346]:
```suggestion
            type Movie @query(aggregate: true) {
```

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 6]
```diff
@@ -0,0 +1,1611 @@
...
+            type UpdateMoviesMutationResponse {
+              info: UpdateInfo!
+              movies: [Movie!]!
+            }"
+        `);
+    });
+
+    test("Disable read fields on subscriptions", async () => {
+        const typeDefs = gql`
+            type Movie {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:498]:
```suggestion
            type Movie @query(aggregate: true) {
```

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 7]
```diff
@@ -0,0 +1,1611 @@
...
+
+    describe("relationships fields", () => {
+        test("Disable read on relationship field", async () => {
+            const typeDefs = gql`
+                type Movie @query(aggregate: true) {
+                    title: String!
+                    description: String
+                }
+
+                type Actor {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:726]:
```suggestion
                type Actor @query(aggregate: true) {
```

File: packages/graphql/tests/schema/directives/selectable.test.ts [hunk 8]
```diff
@@ -0,0 +1,1611 @@
...
+            `);
+        });
+        test("Disable aggregation on relationship field", async () => {
+            const typeDefs = gql`
+                type Movie @query(aggregate: true) {
+                    title: String!
+                    description: String
+                }
+
+                type Actor {
```
Comment (Liam-Doodson) [inline packages/graphql/tests/schema/directives/selectable.test.ts:1163]:
```suggestion
                type Actor @query(aggregate: true) {
```

**Review by Liam-Doodson — APPROVED** [review]
LGTM!

File: packages/graphql/tests/schema/directives/selectable.test.ts (hunk 4 above)
Comment (angrykoala) [reply packages/graphql/tests/schema/directives/selectable.test.ts:193]: you are right, copy-pasting was a bit of a miss here :sweat:

**Pull request merged**